Add FFT extension to the spec (#189)
data-apis/array-api: spec/extensions/fourier_transform_functions.md

# Fourier transform Functions

> Array API specification for Fourier transform functions.

A conforming implementation of the array API standard must provide and support the following functions adhering to the following conventions.

-   Positional parameters must be [positional-only](https://www.python.org/dev/peps/pep-0570/) parameters. Positional-only parameters have no externally-usable name. When a function accepting positional-only parameters is called, positional arguments are mapped to these parameters based solely on their order.
-   Optional parameters must be [keyword-only](https://www.python.org/dev/peps/pep-3102/) arguments.
-   Broadcasting semantics must follow the semantics defined in {ref}`broadcasting`.
-   Unless stated otherwise, functions must support the data types defined in {ref}`data-types`.
-   Unless stated otherwise, functions must adhere to the type promotion rules defined in {ref}`type-promotion`.
-   Unless stated otherwise, floating-point operations must adhere to IEEE 754-2019.

## Objects in API

<!-- NOTE: please keep the functions and their inverse together -->
(function-fft)=
### fft(a, /, *, n=None, axis=-1, norm='backward')

> Computes the one-dimensional discrete Fourier transform. The expected behavior includes a round-trip transform using the inverse function, `fft(ifft(a)) == a`.

Reviewer: It's best to revert the order. For C2C transforms it doesn't matter, but it matters for the R2C->C2R round trip.
```suggestion
Computes the one-dimensional discrete Fourier transform. The expected behavior includes a round-trip transform using the inverse function, `ifft(fft(a)) == a`.
```

#### Parameters

-   **a**: _&lt;array&gt;_

>     -   Input array.

Reviewer 2: We should indicate the required dtypes for inputs and the expected dtypes for the outputs (see other parts of the spec, for instance, elementwise functions, for how this is generally worded).

Reviewer: Aha! @asmeurer good catch. I missed that we need to specify the type casting rules. `numpy.fft` did a bad decision to always return in double precision, which we (CuPy) did not appreciate and deviated on purpose:
https://docs.cupy.dev/en/stable/reference/fft.html#code-compatibility-features
Seriously, with all these shiny backends (pocketfft, FFTW, cuFFT, ...) this is really not necessary.

Author: Should we always expect single precision for the output? Or should it depend on the input dtype?

Reviewer: Sorry I dropped the ball, Stephannie. I'd say for floating-number (real or complex) inputs we keep the same precision, and reject integer inputs as done elsewhere in the standard.

-   **n**: _int_

    -   Length of the transformed axis of the output. If given, the input will be either zero-padded or trimmed to the length `n` before computing the Fourier transform. Otherwise, the length of the input along the axis given by the `axis` keyword. Default: `None`.

-   **axis**: _int_

    -  Axis used to compute the Fourier transform. If it is not specified, the last axis is used. Default: `-1`.

-   **norm**: _str_

    -   Specify the normalization mode. Should be one of the following modes:

        - `'backward'`: No normalization.
        - `'ortho'`: Normalize by `1/sqrt(n)`
        - `'forward'`: Normalize by `1/n`.

        Default: `'backward'`

#### Returns

-   **out**: _&lt;array&gt;_

    -   An array transformed along the axis indicated by the `axis` keyword.

#### Raises

-   If `axis` is not a valid axis of `a`.

(function-ifft)=
### ifft(a, /, *, n=None, axis=-1, norm='backward')

> Computes the one-dimensional inverse discrete Fourier transform.

Reviewer: ```suggestion
Computes the one-dimensional inverse discrete Fourier transform. The expected behavior includes a round-trip transform using the inverse function, `ifft(fft(a)) == a`.
```

Reviewer 2: Do we expect exact floating-point equality? The wording here seems to imply this.

Reviewer: I took it that any equality implies "within numerical accuracy" here 😅 So yes, if this is a concern we should clarify it, as is done in `numpy.fft` docs.

#### Parameters

-   **a**: _&lt;array&gt;_

    -   Input array.

-   **n**: _int_

    -   Length of the transformed axis of the output. If given, the input will be either zero-padded or trimmed to the length `n` before computing the inverse Fourier transform. Otherwise, the length of the input along the axis given by the `axis` keyword. Default: `None`.

-   **axis**: _int_

    -  Axis used to compute the inverse Fourier transform. If it is not specified, the last axis is used. Default: `-1`.

-   **norm**: _str_

    -   Specify the normalization mode. Should be one of the following modes:

        - `'backward'`: Normalize by `1/n`.
        - `'ortho'`: Normalize by `1/sqrt(n)`
        - `'forward'`: No normalization.

        Default: `'backward'`

#### Returns

-   **out**: _&lt;array&gt;_

    -   An array transformed along the axis indicated by the `axis` keyword.

#### Raises

-   If `axis` is not a valid axis of `a`.

(function-fftn)=
### fftn(a, /, *, s=None, axes=None, norm='backward')

> Computes the n-dimensional discrete Fourier transform.

Reviewer: ```suggestion
Computes the n-dimensional discrete Fourier transform. The expected behavior includes a round-trip transform using the inverse function, `ifftn(fftn(a)) == a`.
```

    -   Input array.

-   **s**: _Union\[ Sequence\[ int ], Tuple\[ int, ... ] ]_

    -   Size of each transformed axis of the output. If given, each axis will be either zero-padded or trimmed to the length `s[i]` before computing the Fourier transform. Otherwise, the length of the input along the axis given by the `axes` keyword. Default: `None`.

-   **axes**: _Union\[ Sequence\[ int ], Tuple\[ int, ... ] ]_

    -  Axes over which to compute the Fourier transform. If not specified, the last `len(s)` axes are used, or all axes if `s` is not specified either. Default: `None`.

-   **norm**: _str_

    -   Specify the normalization mode. Should be one of the following modes:

        - `'backward'`: No normalization.
        - `'ortho'`: Normalize by `1/sqrt(n)`
        - `'forward'`: Normalize by `1/n`.

        Default: `'backward'`

#### Returns

-   **out**: _&lt;array&gt;_

    -   An array transformed along the axes indicated by the `axes` keyword.

#### Raises

-   If `s` and `axes` have different lengths.
-   If an element of `axes` is larger than the number of axes of `a`.

(function-ifftn)=
### ifftn(a, /, *, s=None, axes=None, norm='backward')

> Computes the n-dimensional inverse discrete Fourier transform.

Reviewer: ```suggestion
Computes the n-dimensional inverse discrete Fourier transform. The expected behavior includes a round-trip transform using the inverse function, `ifftn(fftn(a)) == a`.
```

    -   Input array.

-   **s**: _Union\[ Sequence\[ int ], Tuple\[ int, ... ] ]_

    -   Size of each transformed axis of the output. If given, each axis will be either zero-padded or trimmed to the length `s[i]` before computing the inverse Fourier transform. Otherwise, the length of the input along the axis given by the `axes` keyword. Default: `None`.

-   **axes**: _Union\[ Sequence\[ int ], Tuple\[ int, ... ] ]_

    -  Axes over which to compute the inverse Fourier transform. If not specified, the last `len(s)` axes are used, or all axes if `s` is not specified either. Default: `None`.

-   **norm**: _str_

    -   Specify the normalization mode. Should be one of the following modes:

        - `'backward'`: Normalize by `1/n`.
        - `'ortho'`: Normalize by `1/sqrt(n)`
        - `'forward'`: No normalization.

        Default: `'backward'`

#### Returns

-   **out**: _&lt;array&gt;_

    -   An array transformed along the axes indicated by the `axes` keyword.

#### Raises

-   If `s` and `axes` have different lengths.
-   If an element of `axes` is larger than the number of axes of `a`.

(function-rfft)=
### rfft(a, /, *, n=None, axis=-1, norm='backward')

> Computes the one-dimensional discrete Fourier transform for real-valued input.

Reviewer: ```suggestion
Computes the one-dimensional discrete Fourier transform for real-valued input. The expected behavior includes a round-trip transform using the inverse function, `irfft(rfft(a), n=a.shape[axis]) == a`.
```

#### Parameters

-   **a**: _&lt;array&gt;_

    -   Input array.

-   **n**: _int_

    -   Length of the input array. If given, the input will be either zero-padded or trimmed to this length before computing the real Fourier transform. Otherwise, the length of the input along the axis specified by the `axis` keyword is used. Default: `None`.

-   **axis**: _int_

    -  Axis used to compute the real Fourier transform. If it is not specified, the last axis is used. Default: `-1`.

-   **norm**: _str_

    -   Specify the normalization mode. Should be one of the following modes:

        - `'backward'`: No normalization.
        - `'ortho'`: Normalize by `1/sqrt(n)`
        - `'forward'`: Normalize by `1/n`.

        Default: `'backward'`

#### Returns

-   **out**: _&lt;array&gt;_

    -   An array transformed along the axis indicated by the `axis` keyword.

#### Raises

-   If `axis` is not a valid axis of `a`.

(function-irfft)=
### irfft(a, /, *, n=None, axis=-1, norm='backward')

> Computes the one-dimensional inverse discrete Fourier transform for real-valued input.

Reviewer: I think "for real-valued input" here is a bit awkward as this is in fact a complex-to-real (C2R) transform...How about this?
```suggestion
Computes the one-dimensional inverse of `rfft`. The expected behavior includes a round-trip transform using the inverse function, `irfft(rfft(a), n=a.shape[axis]) == a`
```

#### Parameters

-   **a**: _&lt;array&gt;_

    -   Input array.

-   **n**: _int_

>     -   Length of the transformed axis of the output. If given, the input will be either zero-padded or trimmed to this length before computing the real Fourier transform. Otherwise, it will default to `2 * (m - 1)` where `m` is the length of the input along the axis given by the `axis` keyword. Default: `None`.

Reviewer: Ditto. Also, when `n` is not None the padding/trimming rule is different for C2R transforms.
```suggestion
    -   Length of the transformed axis of the *output*. If given, the input will be either zero-padded or trimmed to `n//2+1` before computing the inverse of `rfft`. Otherwise, it will default to `2 * (m - 1)` where `m` is the length of the input along the axis given by the `axis` keyword. Default: `None`.
```

-   **axis**: _int_

    -  Axis used to compute the real Fourier transform. If it is not specified, the last axis is used. Default: `-1`.

-   **norm**: _str_

    -   Specify the normalization mode. Should be one of the following modes:

        - `'backward'`: Normalize by `1/n`.
        - `'ortho'`: Normalize by `1/sqrt(n)`
        - `'forward'`: No normalization.

        Default: `'backward'`

#### Returns

-   **out**: _&lt;array&gt;_

>     -   An array transformed along the axis indicated by the `axis` keyword.

Reviewer: ```suggestion
    -   A real-valued array transformed along the axis indicated by the `axis` keyword. The length along the transformed axis is `n` (if given) or `2 * (m - 1)`.
```

#### Raises

-   If `axis` is not a valid axis of `a`.

(function-rfftn)=
### rfftn(a, /, *, s=None, axes=None, norm='backward')

> Computes the n-dimensional discrete Fourier transform for real-valued input.

Reviewer: ```suggestion
Computes the n-dimensional discrete Fourier transform for real-valued input. The expected behavior includes a round-trip transform using the inverse function, `irfftn(rfftn(a), s=a.shape) == a`.
```

#### Parameters

-   **a**: _&lt;array&gt;_

    -   Input array.

-   **s**: _Union\[ Sequence\[ int ], Tuple\[ int, ... ] ]_

>     -   Size of each transformed axis of the output. If given, each axis will be either zero-padded or trimmed to the length `s[i]` before computing the real Fourier transform. Otherwise, no padding will be performed in each dimension. Default: `None`.

Reviewer: ```suggestion
    -   Size of each transformed axis of the output. If given, each axis `i` will be either zero-padded or trimmed to the length `s[i]` before computing the real Fourier transform. Otherwise, the shape of the input along the axes given by the `axes` keyword. The last element `s[-1]` is for computing `rfft(a[axes[-1]], n=s[-1])` whereas other elements for `fft(a[axes[i]], n=s[i])`. Default: `None`.
```

-   **axes**: _Union\[ Sequence\[ int ], Tuple\[ int, ... ] ]_

    -  Axes over which to compute the Fourier transform. If not specified, the last `len(s)` axes are used, or all axes if `s` is not specified either. Default: `None`.

-   **norm**: _str_

    -   Specify the normalization mode. Should be one of the following modes:

        - `'backward'`: No normalization.
        - `'ortho'`: Normalize by `1/sqrt(n)`
        - `'forward'`: Normalize by `1/n`.

        Default: `'backward'`

#### Returns

-   **out**: _&lt;array&gt;_

>     -   An array transformed along the axes indicated by the `axes` keyword.

Reviewer: ```suggestion
    -   A complex-valued array transformed along the axes indicated by the `axes` keyword. The length along the last transformed axis is `s[-1]//2+1` and along other axes `s[i]`.
```

#### Raises

-   If `s` and `axes` have different lengths.

> -   If an element of `axes` is larger than the number of axes of `a`.

Reviewer: ```suggestion
-   If `axes` contains any invalid axis of `a`.
```

(function-irfftn)=
### irfftn(a, /, *, s=None, axes=None, norm='backward')

> Computes the n-dimensional inverse discrete Fourier transform for real-valued input.

Reviewer: ```suggestion
Computes the n-dimensional inverse of `rfftn`. The expected behavior includes a round-trip transform using the inverse function, `irfftn(rfftn(a), s=a.shape) == a`.
```

#### Parameters

-   **a**: _&lt;array&gt;_

    -   Input array.

-   **s**: _Union\[ Sequence\[ int ], Tuple\[ int, ... ] ]_

>     -   Size of each transformed axis of the output. If given, each axis will be either zero-padded or trimmed to the length `s[i]` before computing the Fourier transform. Otherwise, no padding will be performed in each dimension. Default: `None`.

Reviewer: ```suggestion
    -   Size of each transformed axis of the *output*. If given, the last axis will be either zero-padded or trimmed to `s[-1]//2+1`, whereas all other axes `i` are either zero-padded or trimmed to the length `s[i]`, before computing the inverse of `rfftn`. Otherwise, the last axis is either zero-padded or trimmed to `2 * (m - 1)`, where `m` is the length of the input along the axis, and all other axes use the input shape. The last element `s[-1]` is for computing `irfft(a[axes[-1]], n=s[-1])` whereas other elements for `ifft(a[axes[i]], n=s[i])`. Default: `None`.
```

-   **axes**: _Union\[ Sequence\[ int ], Tuple\[ int, ... ] ]_

    -  Axes over which to compute the inverse Fourier transform. If it is not specified, the last `len(s)` axes are used or all axes if `s` is also not specified. Default: `None`.

-   **norm**: _str_

    -   Specify the normalization mode. Should be one of the following modes:

        - `'backward'`: Normalize by `1/n`.
        - `'ortho'`: Normalize by `1/sqrt(n)`
        - `'forward'`: No normalization.

        Default: `'backward'`

#### Returns

-   **out**: _&lt;array&gt;_

>     -   An array transformed along the axes indicated by the `axes` keyword.

Reviewer: ```suggestion
    -   A real-valued array transformed along the axes indicated by the `axes` keyword. The length along the last transformed axis is `s[-1]` (if given) or `2 * (m - 1)`, and all other axes `s[i]`.
```

#### Raises

-   If `s` and `axes` have different lengths.

> -   If an element of `axes` is larger than the number of axes of `a`.

Reviewer: ```suggestion
-   If `axes` contains any invalid axis of `a`.
```

# Fourier transform Functions

> Array API specification for Fourier transform functions.

A conforming implementation of the array API standard must provide and support the following functions adhering to the following conventions.

-   Positional parameters must be [positional-only](https://www.python.org/dev/peps/pep-0570/) parameters. Positional-only parameters have no externally-usable name. When a function accepting positional-only parameters is called, positional arguments are mapped to these parameters based solely on their order.
-   Optional parameters must be [keyword-only](https://www.python.org/dev/peps/pep-3102/) arguments.
-   Broadcasting semantics must follow the semantics defined in {ref}`broadcasting`.
-   Unless stated otherwise, functions must support the data types defined in {ref}`data-types`.
-   Unless stated otherwise, functions must adhere to the type promotion rules defined in {ref}`type-promotion`.

> -   Unless stated otherwise, floating-point operations must adhere to IEEE 754-2019.

Reviewer: Just a note for a potential future PR: Perhaps we want to move this section elsewhere and just refer to it here, instead of copying and pasting in every extension's page?

Reviewer: Oh Ralf was actually ahead of me: https://github.com/data-apis/array-api/pull/189#discussion_r646609536

Reviewer 3: Yes, it's a necessary refactor. I'll open a separate issue for it so we actually do it.

## Objects in API

> <!-- NOTE: please keep the functions in alphabetical order -->

Reviewer: Please don't sort alphabetically...For other modules it may be fine, but for FFT I expect to see `ifft` follows immediately after `fft` and so on, so that we can make a direct comparison between forward and backward transforms. During the review I had to jump back and forth to check things 😢

Author: I'll change this! It was also a little bit weird for me writing the extension alphabetically but I also agree that it is really hard to read

(function-fft)=
### fft(a, /, *, n=None, axis=-1, norm='backward')

> Computes the one-dimensional discrete Fourier transform.

Reviewer: Discussion: Do we want to comment on the expected behavior of a round-trip transform (ex: `fft(ifft(a)) == a`)? I know NumPy does.

Reviewer 3: Seems worth a note.

#### Parameters

-   **a**: _&lt;array&gt;_

    -   Input array.

-   **n**: _int_

>     -   Length of the transformed axis of the output. If it is `None`, the length of the input along the axis given by the `axis` keyword. Default: `None`.

Reviewer: Let's copy `ifft`'s description (which contains additional descriptions like padding/cropping) to here.

-   **axis**: _int_

    -  Axis used to compute the Fourier transform. If it is not specified, the last axis is used. Default: `-1`.

-   **norm**: _str_

    -   Specify the normalization mode. Should be one of the following modes:

        - `'backward'`: No normalization.
        - `'ortho'`: Normalize by `1/sqrt(n)`
        - `'forward'`: Normalize by `1/n`.

        Default: `'backward'`

#### Returns

-   **out**: _&lt;array&gt;_

    -   An array transformed along the axis indicated by the `axis` keyword.

#### Raises

> -   If `axis` is larger than the last axis of `a`.

Reviewer: I found NumPy's description a bit cleaner (excluding both `axis>a.ndim-1` *and* `axis<-a.ndim`).
```suggestion
-   If `axis` is not a valid axis of `a`.
```

(function-fftfreq)=
### fftfreq(n, /, *, d=1.0)

Returns the discrete Fourier transform sample frequencies. For a Fourier transform of length `n` and length unit of `d` the frequencies are described as:

```
f = [0, 1, ..., n/2-1, -n/2, ..., -1] / (d*n) if n is even
f = [0, 1, ..., (n-1)/2, -(n-1)/2, ..., -1] / (d*n) if n is odd
```

#### Parameters

-   **n**: _int_

    -   Window length.

-   **d**: _float_

    -   Sample spacing between individual samples of the Fourier transform input. Default: `1.0`.

#### Returns

-   **out**: _&lt;array&gt;_

    -   An array of length `n` containing the sample frequencies.

(function-fftn)=
### fftn(a, /, *, s=None, axes=None, norm='backward')

Computes the n-dimensional discrete Fourier transform.

#### Parameters

-   **a**: _&lt;array&gt;_

    -   Input array.

-   **s**: _Union\[ Sequence\[ int ], Tuple\[ int, ... ] ]_

>     -   Size of each transformed axis of the output. If given, each axis will be either zero-padded or trimmed to the length `s[i]` before computing the fast Fourier transform. Otherwise, no padding will be performed in each dimension. Default: `None`.

Reviewer: ```suggestion
    -   Size of each transformed axis of the output. If given, each axis will be either zero-padded or trimmed to the length `s[i]` before computing the Fourier transform. Otherwise, the length of the input along the axis given by the `axes` keyword. Default: `None`.
```

-   **axes**: _Union\[ Sequence\[ int ], Tuple\[ int, ... ] ]_

>     -  Axes over which to compute the Fourier transform. If it is not specified, the last `len(s)` axes are used or all axes if `s` is also not specified. Default: `None`.

Reviewer: Do we need a comma?
```suggestion
    -  Axes over which to compute the Fourier transform. If not specified, the last `len(s)` axes are used, or all axes if `s` is not specified either. Default: `None`.
```

-   **norm**: _str_

    -   Specify the normalization mode. Should be one of the following modes:

        - `'backward'`: No normalization.
        - `'ortho'`: Normalize by `1/sqrt(n)`
        - `'forward'`: Normalize by `1/n`.

        Default: `'backward'`

#### Returns

-   **out**: _&lt;array&gt;_

    -   An array transformed along the axes indicated by the `axes` keyword.

#### Raises

> -   If `s` and `axes` have a different length.

Reviewer: ```suggestion
-   If `s` and `axes` have different lengths.
```

> -   If an element of `axes` is larger than the number of axes of `a`.

Reviewer: ditto (see my comment for `fft`)

#### Parameters

-   **a**: _&lt;array&gt;_

    -   Input array.

-   **n**: _int_

    -   Length of the transformed axis of the output. If given, the input will be either zero-padded or trimmed to this length before computing the Hermitian Fourier transform. Otherwise, it will default to `2 * (m - 1)` where `m` is the length of the input along the axis given by the `axis` keyword. Default: `None`.

-   **axis**: _int_

    -  Axis used to compute the Fourier transform. If it is not specified, the last axis is used. Default: `-1`.

-   **norm**: _str_

    -   Specify the normalization mode. Should be one of the following modes:

        - `'backward'`: No normalization.
        - `'ortho'`: Normalize by `1/sqrt(n)`
        - `'forward'`: Normalize by `1/n`.

        Default: `'backward'`

#### Returns

-   **out**: _&lt;array&gt;_

    -   A transformed array.

#### Raises

-   If `axis` is larger than the last axis of `a`.

(function-ifft)=
### ifft(a, /, *, n=None, axis=-1, norm='backward')

Computes the one-dimensional inverse discrete Fourier transform.

#### Parameters

-   **a**: _&lt;array&gt;_

    -   Input array.

-   **n**: _int_

>     -   Length of the transformed axis of the output. If given, the input will be either zero-padded or trimmed to the length `n` before computing the IFourier transform. Otherwise, the length of the input along the axis given by the `axis` keyword. Default: `None`.

Reviewer: ```suggestion
    -   Length of the transformed axis of the output. If given, the input will be either zero-padded or trimmed to the length `n` before computing the inverse Fourier transform. Otherwise, the length of the input along the axis given by the `axis` keyword. Default: `None`.
```

-   **axis**: _int_

>     -  Axis used to compute the Fourier transform. If it is not specified, the last axis is used. Default: `-1`.

Reviewer: ```suggestion
    -  Axis used to compute the inverse Fourier transform. If it is not specified, the last axis is used. Default: `-1`.
```

-   **norm**: _str_

    -   Specify the normalization mode. Should be one of the following modes:

        - `'backward'`: Normalize by `1/n`.
        - `'ortho'`: Normalize by `1/sqrt(n)`
        - `'forward'`: No normalization.

        Default: `'backward'`

#### Returns

-   **out**: _&lt;array&gt;_

    -   An array transformed along the axis indicated by the `axis` keyword.

#### Raises

> -   If `axis` is larger than the last axis of `a`.

Reviewer: ditto

(function-ifftn)=
### ifftn(a, /, *, s=None, axes=None, norm='backward')

Computes the n-dimensional inverse discrete Fourier transform.

#### Parameters

-   **a**: _&lt;array&gt;_

    -   Input array.

-   **s**: _Union\[ Sequence\[ int ], Tuple\[ int, ... ] ]_

>     -   Size of each transformed axis of the output. If given, each axis will be either zero-padded or trimmed to the length `s[i]` before computing the Fourier transform. Otherwise, no padding will be performed in each dimension. Default: `None`.

Reviewer: See my comment for `fftn`. Also, "the Fourier transform" -> "the inverse Fourier transform".

-   **axes**: _Union\[ Sequence\[ int ], Tuple\[ int, ... ] ]_

>     -  Axes over which to compute the inverse Fourier transform. If it is not specified, the last `len(s)` axes are used or all axes if `s` is also not specified. Default: `None`.

Reviewer: See my comment for `fftn`.

-   **norm**: _str_

    -   Specify the normalization mode. Should be one of the following modes:

        - `'backward'`: Normalize by `1/n`.
        - `'ortho'`: Normalize by `1/sqrt(n)`
        - `'forward'`: No normalization.

        Default: `'backward'`

#### Returns

-   **out**: _&lt;array&gt;_

    -   An array transformed along the axes indicated by the `axes` keyword.

#### Raises

-   If `s` and `axes` have a different length.

> -   If an element of `axes` is larger than the number of axes of `a`.

Reviewer: See my comment for `fftn`.
